### PR TITLE
Launch validation 

### PR DESCRIPTION
**Pull request by csweaver**
Added unimplemented validation from https://docs.google.com/document/d/1OaufjWcnAywLg3iQYn_0FltHFUponllXaF9CeQqM0Nc/edit

I, unfortunately, comingled this with a cleanup of imports and double-quoted strings and my cursory attempt to untangle them was taking way too much effort.

I have added a set of bad datasets here for testing purposes: https://drive.google.com/drive/u/1/folders/128v4VA4h2XjScU2R3bDWA_noPqCF75gO

I did not do the not-sparse warning yet. I know @bkmartinjr, you have done some performance benchmarking work so I wanted to get your opinion on if it is still necessary and if so, what is the breakpoint of data size where we should warn if it is not sparse.

### CODE REVIEW

File: server/app/driver/driver.py [hunk 1]
```diff
@@ -101,7 +101,7 @@ def layout(self, filter, interactive_limit=None):
         """
         Computes a n-d layout for cells through dimensionality reduction.
         :param filter: filter: dictionary with filter params
-        :param interactive_limit: -- don't compute if total # genes in dataframes are larger than this
+        :param interactive_limit: -- don"t compute if total # genes in dataframes are larger than this
```
Comment (bkmartinjr) [inline server/app/driver/driver.py]: this should not be a double quote.

File: server/app/scanpy_engine/scanpy_engine.py [hunk 2]
```diff
@@ -40,27 +42,28 @@ def _alias_annotation_names(self, axis, name):
...
             return
 
         ax_name = str(axis)
         df_axis = getattr(self.data, ax_name)
         if name is None:
-            # reset index to simple range; alias 'name' to point at the
+            # reset index to simple range; alias "name" to point at the
             # previously specified index.
-            df_axis = df_axis.reset_index().rename(columns={'index': 'name'})
+            df_axis.reset_index(inplace=True)
```
Comment (bkmartinjr) [inline server/app/scanpy_engine/scanpy_engine.py]: Just want to confirm you have validated the performance implications of using `inplace`?  (I don't know of any issues, just think it is worth checking)

File: server/app/scanpy_engine/scanpy_engine.py [hunk 3]
```diff
@@ -139,14 +154,34 @@ def _validate_data_types(self):
...
+                                "uint32": "int32",
+                                "uint64": "int32",
+                                "float64": "float32",
                                 }
                 if datatype in downcast_map:
                     warnings.warn(f"Scanpy annotation {ax}:{ann} is in unsupported format: {datatype}. "
                                   f"Data will be downcast to {downcast_map[datatype]}.")
+                if isinstance(datatype, CategoricalDtype):
+                    category_num = len(curr_axis[ann].dtype.categories)
+                    if category_num > 100 and category_num > self.max_category_items:
```
Comment (bkmartinjr) [inline server/app/scanpy_engine/scanpy_engine.py]: I think 100 is too conservative (low), and recommend a higher number.   Perhaps 500.   There isn't much difference on the front-end until we get into the 500-1000 range, so doesn't seem worth the warning at 100.

File: server/cli/launch.py [hunk 4]
```diff
@@ -76,24 +77,32 @@ def launch(data, layout, diffexp, title, verbose, debug, obs_names, var_names,
...
+        click.echo(f"[cellxgene] Launching! Opening your browser to {cellxgene_url} now.")
         webbrowser.open(cellxgene_url)
     else:
-        click.echo(f'[cellxgene] Launching! Please go to {cellxgene_url} in your browser.')
+        click.echo(f"[cellxgene] Launching! Please go to {cellxgene_url} in your browser.")
 
-    click.echo('[cellxgene] Type CTRL-C at any time to exit.')
+    click.echo("[cellxgene] Type CTRL-C at any time to exit.")
 
     app.run(host=host, debug=debug, port=port)
```
Comment (bkmartinjr) [inline server/cli/launch.py]: what happened to threading=true?   (is this branch merged with latest master?)

File: server/app/scanpy_engine/scanpy_engine.py (hunk 3 above)
Comment (csweaver) [reply server/app/scanpy_engine/scanpy_engine.py]: ✅

File: server/app/driver/driver.py (hunk 1 above)
Comment (csweaver) [reply server/app/driver/driver.py]: 🤦‍♀️

File: server/cli/launch.py (hunk 4 above)
Comment (csweaver) [reply server/cli/launch.py]: Hmm, I thought so. I'll double check

File: server/app/scanpy_engine/scanpy_engine.py (hunk 2 above)
Comment (csweaver) [reply server/app/scanpy_engine/scanpy_engine.py]: Speeds it up actually:
Master: TIME 0.2940518856048584
This Branch: TIME: 0.04870891571044922

Comment (csweaver) [reply server/app/scanpy_engine/scanpy_engine.py]: ⬆️ is for the 1.3 mil dataset

File: server/cli/launch.py (hunk 4 above)
Comment (csweaver) [reply server/cli/launch.py]: rebased

File: server/app/scanpy_engine/scanpy_engine.py (hunk 2 above)
Comment (bkmartinjr) [reply server/app/scanpy_engine/scanpy_engine.py]: nice!

**Pull request merged**